crypto-primitives/src/sponge/rescue/constraints.rs

```diff
@@ -0,0 +1,328 @@
+use crate::sponge::constraints::AbsorbGadget;
+use crate::sponge::constraints::{CryptographicSpongeVar, SpongeWithGadget};
+use crate::sponge::rescue::{RescueConfig, RescueSponge};
+use crate::sponge::DuplexSpongeMode;
+use ark_ff::PrimeField;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_r1cs_std::prelude::*;
+use ark_relations::gr1cs::{ConstraintSystemRef, SynthesisError};
+
+#[cfg(not(feature = "std"))]
+use ark_std::vec::Vec;
+
+#[derive(Clone)]
+/// the gadget for Rescue sponge
+///
+/// This implementation of Rescue is entirely from Fractal's implementation in [COS20][cos]
+/// with small syntax changes.
+///
+/// [cos]: https://eprint.iacr.org/2019/1076
+pub struct RescueSpongeVar<F: PrimeField> {
+    /// Constraint system
+    pub cs: ConstraintSystemRef<F>,
+
+    /// Sponge Parameters
+    pub parameters: RescueConfig<F>,
+
+    // Sponge State
+    /// The sponge's state
+    pub state: Vec<FpVar<F>>,
+    /// The mode
+    pub mode: DuplexSpongeMode,
+}
+
+impl<F: PrimeField> SpongeWithGadget<F> for RescueSponge<F> {
+    type Var = RescueSpongeVar<F>;
+}
+
+impl<F: PrimeField> RescueSpongeVar<F> {
+    #[cfg(all(feature = "gr1cs", not(feature = "r1cs")))]
+    fn apply_s_box(
+        &self,
+        state: &mut [FpVar<F>],
+        exponent: &[u64],
+        alpha: u64,
+        // round: usize
+    ) -> Result<(), SynthesisError> {
+        use ark_relations::lc;
+        let cs = state[0].cs();
+
+        if [alpha] == exponent {
+            for state_item in state.iter_mut() {
+                let new_state_item =
+                    FpVar::new_witness(self.cs(), || Ok(state_item.value()?.pow(exponent)))
+                        .unwrap();
+                match (&state_item, &new_state_item) {
+                    (FpVar::Var(alloc_fp), FpVar::Var(new_alloc_fp)) => {
+                        let _ = cs.enforce_constraint(
+                            "XXX",
+                            vec![lc!() + alloc_fp.variable, lc!() + new_alloc_fp.variable],
+                        );
+                        *state_item = new_state_item;
+                    }
+                    _ => {
+                        *state_item = state_item.pow_by_constant(exponent)?;
+                    }
+                }
+            }
+        } else {
+            for state_item in state.iter_mut() {
+                let new_state_item =
+                    FpVar::new_witness(self.cs(), || Ok(state_item.value()?.pow(exponent)))
+                        .unwrap();
+                match (&state_item, &new_state_item) {
+                    (FpVar::Var(alloc_fp), FpVar::Var(new_alloc_fp)) => {
+                        let _ = cs.enforce_constraint(
+                            "XXX",
```

Comment:
This should be "Deg5Mul". Fix in relations to scan the list of predicates and find the R1CS predicate.